working/0546-patterns/records-feature-specification.md

```diff
@@ -283,6 +283,41 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression is a compile-time constant expression
+if all its record field expressions are compile-time constant expressions. 
+
+_This is true whether the expression occurs in a constant context or not,
+which means that a record expression can be used directly as a parameter default value 
+if its record field expressions are constant expressions.
+Example: `f({(int, int) x = (1, 2)}) => ...`._
+
+A record expression is a potentially constant expression 
+if all its record field expressions are potentially constant or constant expressions.
+
+_This means that a record expression can be used in the initializer list
+of a constant non-redirecting generative constructor._
+
+Constant *object* instantiations create deeply immutable and canonicalied objects.
+Records are always unmodifiable, and if their field values are deeply immutable,
+like constants values, the records are also deeply immutable.
+It's meaningless to consider whether record constants are canonicalized,
+since records do not have a persistent identity.
+
+Because of that, there is no need for a `const (1, 2)` syntax to force a record 
+to be a constant, like there is for object creation expressions. 
+A record expression with field values that are constant-created values, 
+will be indistinguishable from a similar expression created in a constant 
+context, since identity cannot be used as a distinguishing trait.
```

Comment:
Identity cannot be used as a _negatively_ distinguishing trait, that is, we can't conclude non-sameness from `identical(r1, r2) == false`.

However, this doesn't imply that `identical(r1, r2) == true` is useless, and we might be able to use explicitly constant records to ensure that this occurs more frequently.

Reply:
I really, really hope people *won't* start trying to force canonicalization and depend on the identity of records for anything.
It's an optimization. If your code can't run well if `identical` returns `false` on all records, you should fix it. Or stop using records for something where identity matters.

Reply:
> I really, really hope people won't start trying to force canonicalization and
> depend on the identity of records for anything.

Right, we might very well want a lint against statically known records as arguments to `identical`, except for the use of exactly this approach in the implicitly induced implementation of `==` on record types.

But the point is that an explicit request for canonicalization could ensure that `==` runs faster, more frequently.

Reply:
I'm strongly opposed to giving people the idea that they can affect canonicalization of records, and that it matters in any way.

You really cannot know that canonicalization will make anything run faster. It might. It's definitely true that more canonicalized objects gives more chances of a positive `identical` check.

But you also cannot know that *requesting* canonicalization actually will cause more canonicalization. It's perfectly reasonable for a compiler to canonicalize all the constant records expressions it can see anyway, and if the compiler chooses not to, perhaps it's for a good reason. And it still can choose not to, even if you request canonicalization.

If I was a compiler writer implementing records, I'd completely ignore any user request for canonicalization of records, because I can, and do whatever I think is best for the compilation.

So why give the user a button to press, if it's not connected to anything anyway?

Reply:
Even if the compiler *does* canonicalize every single record, as soon as you *mention* it, the compiler could silently copy it to a new unboxed storage and you've lost the fact that it once referred to some canonical instance.